src/Architecture.cpp

```diff
@@ -185,20 +189,43 @@ Architecture::Architecture(const std::uint16_t nQ, const CouplingMap& cm,
 }
 
 void Architecture::createDistanceTable() {
+  Matrix<double> edgeWeights(nqubits, std::vector<double>(nqubits, INFINITY));
   for (const auto& edge : couplingMap) {
+    edgeWeights.at(edge.first).at(edge.second) = COST_BIDIRECTIONAL_SWAP;
     if (couplingMap.find({edge.second, edge.first}) == couplingMap.end()) {
-      isBidirectional = false;
-      break;
+      edgeWeights.at(edge.second).at(edge.first) = COST_UNIDIRECTIONAL_SWAP;
     }
   }
 
-  if (isBidirectional) {
-    Dijkstra::buildTable(nqubits, couplingMap, distanceTable,
-                         Architecture::costHeuristicBidirectional);
-  } else {
-    Dijkstra::buildTable(nqubits, couplingMap, distanceTable,
-                         Architecture::costHeuristicUnidirectional);
+  Dijkstra::buildTable(nqubits, couplingMap, distanceTable, edgeWeights,
+                       Architecture::dijkstraNodeToCostNonFidelity);
+}
+
+void Architecture::createFidelityDistanceTable() {
+  for (std::uint16_t i = 0; i < nqubits; ++i) {
+    for (std::uint16_t j = 0; j < nqubits; ++j) {
+      if (!properties.twoQubitErrorRateAvailable(i, j)) {
+        fidelityDistanceTable = std::vector<std::vector<double>>(
+            nqubits, std::vector<double>(nqubits, 0));
+        return;
+      }
+    }
   }
+
+  Matrix<double> edgeWeights(nqubits, std::vector<double>(nqubits, INFINITY));
+  for (const auto& edge : couplingMap) {
+    edgeWeights.at(edge.first).at(edge.second) =
+        -3 * log2(fidelityTable.at(edge.first).at(edge.second));
+    if (couplingMap.find({edge.second, edge.first}) == couplingMap.end()) {
+      edgeWeights.at(edge.second).at(edge.first) =
+          -3 * log2(fidelityTable.at(edge.first).at(edge.second)) -
+          2 * log2(singleQubitFidelities.at(edge.first)) -
+          2 * log2(singleQubitFidelities.at(edge.second));
+    }
```

Comment:
I have seen Code like this somewhere before up in the code. Maybe also something to extract and reuse.